docs/src/proposals/interest-bearing-tokens.md

```diff
@@ -0,0 +1,284 @@
+simply need to call the right program to perform instructions.
+
+## Required Changes
+
+A new interest-bearing token program is only one component of the overall solution,
+since i-tokens need to be easily integrated in on-chain programs, web apps,
+and wallets.
+
+### Token Program Conformance
+
+Since our solution entails the creation of a new token program, we're opening the
+floodgates for other app developers to do the same. To ensure safe and easy integration
+of new token programs, we need a set of conformance tests based on the current
+SPL token test suite.
+
+A new token program must include all of the same instructions as SPL token,
+but the "unchecked" variants should return errors. Concretely, this means the
+program must implement:
+
+* `InitializeMint`
+* `InitializeAccount`
+* `InitializeMultisig` -- TODO deprecate this and use a more general multisig?
+* `Revoke`
+* `SetAuthority`
+* `CloseAccount`
+* `FreezeAccount`
+* `ThawAccount`
+* `TransferChecked`
+* `ApproveChecked`
+* `MintToChecked`
+* `BurnChecked`
+* `InitializeAccount2`
+
+And the program should throw errors for:
+
+* `Transfer`
+* `Approve`
+* `MintTo`
+* `Burn`
+
+New instructions required:
+
+* `CreateHolding` (TODO name): only performs `Allocate` and `Assign` to self,
+useful for creating a holding when you don't know the program that you're
+interacting with. See the Associated Token Account section for how to use this.
+
+There are also new read-only instructions, which write data to a provided account
+buffer:
+
+* `GetHoldingBalance`: balance of the holding
+* `GetHoldingOwner`: owner's public key
+* `GetMintSupply`: mint supply
+* `GetMintAuthority`: mint authority's public key
+* `GetMintDecimals`: mint decimals
+
+See the Runtime section for more information on how these are used.
+
+Programs and wallets that wish to support i-tokens must update the instructions
+to use all `Checked` variants and use the new read-only instructions to fetch
+information.
+
+The SPL token library provides wrappers compatible with any conforming program.
+For the base SPL token program, the wrapper simply deserializes the account data
+and returns the relevant data. For other programs, it calls the appropriate
+read-only instruction using an ephemeral account input.
+
+The structure for new holding and mint accounts must follow the layout of the
+existing SPL token accounts to ensure compatibility in RPC when fetching pre and
+post token balances.
+
+For a holding, the first 165 bytes must contain the same information
+as a normal SPL token holding. The following byte must be the type of
+account (ie. `Holding`), and after that, any data is allowed as required by the
+new token program.
+
+The same applies for mints, but only for the first 82 bytes.
+
+The `amount` field in an SPL token holding should contain a balance in order
+to properly integrate with RPC's `preTokenBalances` and `postTokenBalances`.
```

Comment:
`amount` here as in `share`'s amount? Ultimately, `{pre,post}TokenBalances` holds enough information to reconstruct the delta of balances as in `token`'s amount at the time?

Reply:
I'll clarify this part too -- `amount` would actually be the token amount in this case, and not the `share` amount.  Otherwise, I think RPC would need to perform a calculation in order to convert to "UI Amount".  Unless, are you suggesting that the read-only `GetBalance` instruction is run here as well?

Reply:
> are you suggesting that the read-only GetBalance instruction is run here as well

Yeah, I was implying that while worrying about the computing cost. Otherwise, the method would be of less merit... That's because RPC consumers always call another call to rpc? I don't think they can tolerate to display share's amount (internal thing) to end users.

Reply:
Yeah, that's fair.  I think we should keep caching the token amount somewhere as an implementation detail.  This might also help out with ledger support, but I'll discuss that with @t-nelson

Reply:
I think `amount` needs to be shares instead of balance, because all token instructions still take shares as input.  The `uiAmountString` should be the current real balance though

Reply:
> Yeah, that's fair. I think we should keep caching the token amount somewhere as an implementation detail. This might also help out with ledger support, but I'll discuss that with @t-nelson

I wouldn't do any extra work account-side the the Leger app.  All it gets is a serialized transaction message, no other supplemental info like this.

Reply:
> 

Hmm, I thought `amount` of all token/i-token instructions are `balance-amount`/`token-amount`, not `share-amount`. That's because it would be easier for consumers of i-token (end-user, devs) to use in that way?

Say, defi apps want to do `do something() if i-token is more than NN <TICKER>`. Then, if the instruction takes in terms of shares. we need to provide to convert back the value returned from getHoldingBalance into share's value?

I think atoken takes amount in token unit, not share unit:

https://github.com/aave/protocol-v2/blob/3d3dfc01bad46a1b4df6f4bdf47fd8845df848b0/contracts/protocol/tokenization/AToken.sol#L373-L384

Reply:
I was imagining `getHoldingBalance` would return the token unit only.  To get the share unit, one decodes the token holding account like what's done today to read the SPL Token account balance

Reply:
It's not clear to me whether you are all talking about the same `amount`: the `amount` stored within the first 165 token-holding state? or `amount` passed in an instruction/returned from a read-only instruction request?
Based on my understanding of the proposal so far, I assumed the former would be in terms of shares, the latter in terms of tokens.

> are you suggesting that the read-only `GetBalance` instruction is run here as well?

Seems to me like this is necessary. I am also envisioning what mvines just described ^

Reply:
I added a "Proposed Layout" section that might help clarify where my thoughts are on this, and clarified "token amount" vs "share amount".